pkg/util/errors.go

```diff
@@ -4,10 +4,63 @@ import (
 	"bytes"
 	"fmt"
 
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/go-kit/kit/log/level"
+	"github.com/pkg/errors"
 	"google.golang.org/grpc/codes"
 	"google.golang.org/grpc/status"
 )
 
+type CodedError interface {
+	error
+	Code() int
+}
+
+type codedError struct {
+	code    int
+	message string
+}
+
+func NewCodedError(code int, message string) CodedError {
+	return codedError{
+		code:    code,
+		message: message,
+	}
+}
```

Comment:
I'm not clear on why we need this over the existing `httpgrpc` package. It seems like that already does everything we want it to and it can even be retain the HTTP result code over the RPC responses.

Reply:
I think the `CodedErrorSampler` should be in its own package. (`error` package ?)
May be AddEntry should only accept `CodedError`.

We can also add conversion method to convert httpgrpc to create new error.

`FromErr(err,400)` where 400 is the default if no httpgrpc was found while casting otherwise convert it.
`StatusErrorF(status,tmpl,args...)`
`StatusError(status,string)`

Reply:
or even AddFromErr(err,400), Add(400,"error: %v",details)

Reply:
you might need also a helper to convert back to `httpgrpc`